From 57a81d0c90fc509fd197b30378cc0ada3b7afbf1 Mon Sep 17 00:00:00 2001
From: Martin Briza <m@rtinbriza.cz>
Date: Fri, 12 Jul 2013 15:37:05 +0200
Subject: [PATCH 4/4] Fixed GCancellable handling in PolkitQtListener

There was a race condition happening on two simultanneous requests to the agent, causing it to crash.

BUGS: 270489
---
 agent/polkitqtlistener.cpp | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/agent/polkitqtlistener.cpp b/agent/polkitqtlistener.cpp
index a9a93aa..dc5fdda 100644
--- a/agent/polkitqtlistener.cpp
+++ b/agent/polkitqtlistener.cpp
@@ -115,6 +115,13 @@ static void polkit_qt_listener_initiate_authentication(PolkitAgentListener  *age
     qDebug() << "Listener adapter polkit_qt_listener_initiate_authentication";
     PolkitQtListener *listener = POLKIT_QT_LISTENER(agent_listener);
 
+    if (cancellable != NULL) {
+        g_cancellable_connect(cancellable,
+                              G_CALLBACK(cancelled_cb),
+                              agent_listener,
+                              NULL);
+    }
+
     // The result of asynchronous method will be created here and it will be pushed to the listener.
     GSimpleAsyncResult *result = g_simple_async_result_new((GObject *) listener, callback, user_data, agent_listener);
     qDebug() << "GSimpleAsyncResult:" << result;
@@ -129,13 +136,6 @@ static void polkit_qt_listener_initiate_authentication(PolkitAgentListener  *age
             cancellable,
             result);
 
-    if (cancellable != NULL) {
-        g_signal_connect(cancellable,
-                         "cancelled",
-                         G_CALLBACK(cancelled_cb),
-                         agent_listener);
-    }
-
 }
 
 static gboolean polkit_qt_listener_initiate_authentication_finish(PolkitAgentListener  *listener,
-- 
1.8.4.2

